src/main/java/formflow/library/FileController.java

```diff
@@ -181,4 +213,117 @@ RedirectView delete(
+   * that includes the file.
+   * <p>ON FAILURE: A ResponseEntity with a status of forbidden-403 or not found- 404</p>
+   */
+  @GetMapping("/file-download/{submissionId}/{fileId}")
+  public ResponseEntity<StreamingResponseBody> downloadSingleFile(
+      HttpSession httpSession,
+      @PathVariable String submissionId,
+      @PathVariable String fileId
+  ) {
+
+    Optional<UserFile> maybeFile = uploadedFileRepositoryService.findById(UUID.fromString(fileId));
+    if (maybeFile.isEmpty()) {
+      log.error(String.format("File with id %s may have already been deleted", fileId));
+      return ResponseEntity.notFound().build();
+    }
+
+    UserFile file = maybeFile.get();
+    if (!submissionId.equals(httpSession.getAttribute("id").toString())) {
+      return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+    }
+
+    if (!httpSession.getAttribute("id").toString().equals(file.getSubmissionId().getId().toString())) {
+      log.error(String.format("Attempt to download file with submission ID %s but session ID %s does not match",
+          file.getSubmissionId().getId(), httpSession.getAttribute("id")));
+      return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+    }
+
+    String repositoryPath = file.getRepositoryPath();
+    String filename = file.getOriginalName();
+    String contentType = file.getMimeType();
+
+    CloudFile cloudFile = cloudFileRepository.get(repositoryPath);
+    byte[] fileData = cloudFile.getFileBytes();
+    long fileSize = cloudFile.getFileSize();
+
+    StreamingResponseBody responseBody = outputStream -> {
+      try {
+        outputStream.write(fileData);
+      } catch (IOException e) {
+        log.error("Error occurred while downloading file " + e.getMessage());
+      }
+    };
+
+    return ResponseEntity.ok()
+        .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + filename + "\"")
+        .contentType(MediaType.parseMediaType(contentType))
+        .contentLength(fileSize)
+        .body(responseBody);
+  }
+
+  /**
+   * @param httpSession  The current httpsession.
+   * @param submissionId The submissionId of the all the files that you would like to download.
+   * @return ON SUCCESS: A ResponseEntity with a status of ok, a header, and a responseBody that a zip folder with all the files
+   * assoicated with a submission.
+   * <p>ON FAILURE: A ResponseEntity with a status of forbidden-403 or not found- 404</p>
+   */
+  @GetMapping("/file-download/{submissionId}")
+  ResponseEntity<StreamingResponseBody> downloadAllFiles(
+      HttpSession httpSession,
+      @PathVariable String submissionId
+  ) {
+    Optional<Submission> maybeSubmission = submissionRepositoryService.findById(UUID.fromString(submissionId));
+    if (httpSession.getAttribute("id").toString().equals(submissionId) && maybeSubmission.isPresent()) {
+      if (maybeSubmission.isEmpty()) {
```

Comment:
Can it pass "isPresent()" but be empty??  This "if" should probably be the "else" to line 285.  If you get here, there is a submission, as you just checked "isPresent()"